controller/persistence/eventual_eventer.go

```diff
@@ -0,0 +1,607 @@
+		ctx := boltz.NewMutateContext(tx)
+		return a.store.DeleteById(ctx, id)
+	})
+
+	if err == nil {
+		total := atomic.AddInt64(a.outstandingEvents, -1)
+		a.Emit(EventualEventRemovedName, &EventualEventRemoved{
+			Id:    id,
+			Total: total,
+		})
+	}
+
+	return err
+}
+
+// notifyWaiters closes all signaling channels returned in calls
+// to Trigger().
+func (a *EventualEventerBbolt) notifyWaiters() {
+	var waiterIds []string
+
+	a.waiters.Range(func(key, value interface{}) bool {
+		id := key.(string)
+		waiterChan := value.(chan struct{})
+		if waiterChan != nil {
+			close(waiterChan)
+		}
+
+		waiterIds = append(waiterIds, id)
+		return true
+	})
+
+	for _, id := range waiterIds {
+		a.waiters.Delete(id)
+	}
+}
+
+// getEventualEvents returns eventual events up from the persistent bbolt store. The number
+// returned is determined by batchSize and the order of the events are based on key sorting
+func (a *EventualEventerBbolt) getEventualEvents() ([]string, []*EventualEvent, error) {
+	var ids []string
+	var eventualEvents []*EventualEvent
+	err := a.dbProvider.GetDb().View(func(tx *bbolt.Tx) error {
+		var err error
+		ids, _, err = a.store.QueryIds(tx, fmt.Sprintf("limit %d", a.batchSize))
+
+		if err != nil {
+			return err
+		}
+
+		for _, id := range ids {
+			event, err := a.store.LoadOneById(tx, id)
+
+			if err != nil {
+				pfxlog.Logger().WithField("id", id).WithError(err).Errorf("error could not load event id %s", id)
+			} else {
+				eventualEvents = append(eventualEvents, event)
+			}
+		}
+		return nil
+	})
+
+	return ids, eventualEvents, err
+}
+
+// process is the main execution look for dealing with batches of eventual events, triggering listeners for
+// the eventual events and emitting normal processing events
+func (a *EventualEventerBbolt) process() {
+	processId := cuid.New()
+	totalBatches := int64(0)
+	totalEvents := int64(0)
+	totalListenersExecuted := int64(0)
+	totalEventIndex := int64(0)
+
+	a.Emit(EventualEventProcessingStartName, &EventualEventProcessingStart{
+		Id: processId,
+	})
+
+	defer func() {
+		a.notifyWaiters()
+
+		a.Emit(EventualEventProcessingDoneName, &EventualEventProcessingDone{
+			Id:                     processId,
+			TotalBatches:           totalBatches,
+			TotalEvents:            totalEvents,
+			TotalListenersExecuted: totalListenersExecuted,
+		})
+	}()
+
+	for {
+		batchId := cuid.New()
+
+		var numIds = 0
+		var ids []string
+		var eventualEvents []*EventualEvent
+
+		// if we know we have outstanding events query for them up to 5 times then give up
+		numEvents := atomic.LoadInt64(a.outstandingEvents)
+		for attempt := 0; attempt < 5 && numEvents > 0; attempt++ {
+			var err error
+			ids, eventualEvents, err = a.getEventualEvents()
+			numIds = len(ids)
+
+			if err != nil {
+				pfxlog.Logger().WithError(err).Error("error during EventualEventerBbolt processing, could not query events")
+			}
+
+			if numIds > 0 {
+				break
+			}
+
+			time.Sleep(500 * time.Microsecond)
```

Comment:
eh...  so this for loop could block for up to 2.5s? there's only one running right so shouldn't be a big deal? maybe a comment in here?

Reply:
This is dealing with a situation where we know events have been added but the bbolt transaction that added them hasn't been completed yet.

There is a comment, at the top of the loop.

Reply:
It is 500 microseconds, not milliseconds. At most, it will wait 2,500 microseconds aka 2.5 milliseconds.

Reply:
Moved to a separate named function w/ comment.